pop-api/examples/fungibles/tests.rs

```diff
@@ -0,0 +1,831 @@
+			],
+			None,
+		)
+	}
+
+	pub(super) fn approve(
+		session: &mut Session<Sandbox>,
+		spender: AccountId32,
+		value: Balance,
+	) -> Result<(), PSP22Error> {
+		call::<()>(session, "PSP22::approve", vec![spender.to_string(), value.to_string()], None)
+	}
+
+	pub(super) fn increase_allowance(
+		session: &mut Session<Sandbox>,
+		spender: AccountId32,
+		value: Balance,
+	) -> Result<(), PSP22Error> {
+		call::<()>(
+			session,
+			"PSP22::increase_allowance",
+			vec![spender.to_string(), value.to_string()],
+			None,
+		)
+	}
+
+	pub(super) fn decrease_allowance(
+		session: &mut Session<Sandbox>,
+		spender: AccountId32,
+		value: Balance,
+	) -> Result<(), PSP22Error> {
+		call::<()>(
+			session,
+			"PSP22::decrease_allowance",
+			vec![spender.to_string(), value.to_string()],
+			None,
+		)
+	}
+
+	/// PSP22Metadata test utilities.
+
+	pub(super) fn token_name(session: &mut Session<Sandbox>) -> Option<String> {
+		call::<Option<String>>(session, "PSP22Metadata::token_name", vec![], None).unwrap()
+	}
+
+	pub(super) fn token_symbol(session: &mut Session<Sandbox>) -> Option<String> {
+		call::<Option<String>>(session, "PSP22Metadata::token_symbol", vec![], None).unwrap()
+	}
+
+	pub(super) fn token_decimals(session: &mut Session<Sandbox>) -> u8 {
+		call::<u8>(session, "PSP22Metadata::token_decimals", vec![], None).unwrap()
+	}
+
+	/// PSP22Mintable test utilities.
+
+	pub(super) fn mint(
+		session: &mut Session<Sandbox>,
+		account: AccountId32,
+		amount: Balance,
+	) -> Result<(), PSP22Error> {
+		call::<()>(
+			session,
+			"PSP22Mintable::mint",
+			vec![account.to_string(), amount.to_string()],
+			None,
+		)
+	}
+
+	/// PSP22Burnable test utilities.
+
+	pub(super) fn burn(
+		session: &mut Session<Sandbox>,
+		account: AccountId32,
+		amount: Balance,
+	) -> Result<(), PSP22Error> {
+		call::<()>(
+			session,
+			"PSP22Burnable::burn",
+			vec![account.to_string(), amount.to_string()],
+			None,
+		)
+	}
+
+	/// Call a contract method and decode the returned data.
+	pub(super) fn call<T: Decode>(
+		session: &mut Session<Sandbox>,
+		func_name: &str,
+		input: Vec<String>,
+		endowment: Option<Balance>,
+	) -> Result<T, PSP22Error> {
+		match session.call::<String, ()>(func_name, &input, endowment) {
+			// If the call is reverted, decode the error into `PSP22Error`.
+			Err(SessionError::CallReverted(error)) => {
+				Err(PSP22Error::decode(&mut &error[2..])
+					.unwrap_or_else(|_| panic!("Decoding failed")))
+			},
+			// If the call is successful, decode the last returned value.
+			Ok(_) => Ok(session
+				.record()
+				.last_call_return_decoded::<T>()
+				.unwrap_or_else(|_| panic!("Expected a return value"))
+				.unwrap_or_else(|_| panic!("Decoding failed"))),
+			// Catch-all for unexpected results.
+			_ => panic!("Expected call to revert or return a value"),
+		}
+	}
+
+	/// Convert into `PSP22Error::Custom` error type.
+	pub(super) fn into_psp22_custom<T: Encode>(err: T) -> PSP22Error {
```

Comment:
Can we move this to pop-drink as something that converts into a status code. The code from a status code to custom error is then much simpler. We should also define the conversion from a u32 to PSP22Error wherever that type is defined via From, so that the test code can just use .into()

Reply:
This function should simply expect a pop api Error (not generic), convert it into a statuscode, and then in a string.

This should not live in the pop-drink code but in the pop api code. Reason being is that when a dev implements their own contract error, a function that converts a pop api error into a variant of that custom error is necessary for writing tests. This helper function should be placed where the error is defined (i.e. the pop api)

Reply:
This function is removed in #324